File: src/main/ebay/browse.ts
/** eBay Browse API - market price research for a product. */

import { buildPriceStats } from '@main/core/pricing'
import type { PriceStats } from '@shared/types'

import type { EbayClient } from './client'

const SAMPLE_LIMIT = 50

interface ItemSummary {
  price?: { value?: string; currency?: string; convertedFromValue?: string }
}

/**
 * Collect current asking prices for comparable live listings.
 *
 * Prefers an exact GTIN match; falls back to a keyword search. Note these
 * are *asking* prices - actual sold prices need the Marketplace Insights
 * API (separate eBay approval), which can be plugged in here later
 * without changing anything downstream.
 */
export async function priceResearch(
  client: EbayClient,
  options: { gtin?: string | null; query?: string | null },
): Promise<PriceStats | null> {
  const params: Record<string, string | number> = { limit: SAMPLE_LIMIT }
  if (options.gtin) params.gtin = options.gtin
  else if (options.query) params.q = options.query
  else return null

  const data = await client.get<{ itemSummaries?: ItemSummary[] }>(
    '/buy/browse/v1/item_summary/search',
    { params, token: 'app' },
  )

  const currency = client.config.currency
  const prices: number[] = []
  for (const item of data.itemSummaries ?? []) {
    // Skip foreign-currency listings to keep the statistics comparable.
    if (item.price?.convertedFromValue) continue
    if (item.price?.currency !== currency) continue

    const value = Number(item.price.value)
    if (Number.isFinite(value) && value > 0) prices.push(value)
  }

  return prices.length > 0 ? buildPriceStats(prices, currency) : null
}

File: src/epay_tool/ebay/browse.py
"""eBay Browse API - market price research for a product."""

from __future__ import annotations

from epay_tool.core.pricing import PriceStats
from epay_tool.ebay.client import EbayClient


def price_research(
    client: EbayClient, gtin: str | None = None, query: str | None = None
) -> PriceStats | None:
    """Collect current asking prices for comparable live listings.

    Prefers an exact GTIN match; falls back to a keyword search. Note these
    are *asking* prices - actual sold prices need the Marketplace Insights
    API (separate eBay approval), which can be plugged in here later.
    """
    params: dict = {"limit": 50}
    if gtin:
        params["gtin"] = gtin
    elif query:
        params["q"] = query
    else:
        return None

    data = client.get("/buy/browse/v1/item_summary/search", params=params, token="app")
    prices: list[float] = []
    currency = client.settings.currency
    for item in data.get("itemSummaries") or []:
        price = item.get("price") or {}
        if price.get("convertedFromValue"):
            continue  # skip foreign-currency listings to keep stats clean
        try:
            value = float(price["value"])
        except (KeyError, TypeError, ValueError):
            continue
        if price.get("currency") == currency:
            prices.append(value)

    if not prices:
        return None
    return PriceStats.from_prices(prices, currency)
